File: Programmers/[1차] 뉴스 클러스트링.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void to_lowercase(char *str) {
    while (*str) {
        *str = tolower(*str);
        str++;
    }
}

int create_bigrams(char *str, char bigrams[][3], int *count) {
    int len = strlen(str);
    *count = 0;

    for (int i = 0; i < len - 1; i++) {
        if (isalpha(str[i]) && isalpha(str[i + 1])) {
            bigrams[*count][0] = str[i];
            bigrams[*count][1] = str[i + 1];
            bigrams[*count][2] = '\0';
            (*count)++;
        }
    }
    return *count;
}

int solution(char *str1, char *str2) {
    int answer = 0;
    char A[1000][3];
    char B[1000][3];
    int countA = 0, countB = 0;
    int hap, kyo = 0;

    to_lowercase(str1);
    to_lowercase(str2);

    create_bigrams(str1, A, &countA);
    create_bigrams(str2, B, &countB);

    if (countA == 0 && countB == 0) {
        return 65536;
    }

    hap = countA + countB;

    for (int i = 0; i < countA; i++) {
        for (int j = 0; j < countB; j++) {
            if (strcmp(A[i], B[j]) == 0) {
                kyo++;
                for (int k = j; k < countB - 1; k++) {
                    strcpy(B[k], B[k + 1]);
                }
                countB--;
                hap--;
                break;
            }
        }
    }

    answer = (int)((double)kyo / hap * 65536);

    return answer;
}

File: Programmers/[1차] 뉴스 클러스트링.py
# 2018 KAKAO BLIND RECRUITMENT
# 문자열은 대소문자를 신경쓰지 않으므로 소문자로 모두 바꿔준다.
# 알파벳 두 글자씩 끊어서 비교를 하므로, 두 글자씩 끊었을 때에 알파벳인지 아닌지를 판단한다.
# 두 집합이 모두 영어가 없는 경우에 나눌 수 없으므로, 문제에서 주어진대로 65536을 반환한다.
# 두 집합을 이중 for문으로 탐색하면서, 동일한 값이 있으면 교집합+1과 합집합-1을 해준다.
# 교집합/합집합의 값을 반환한다.

def solution(str1, str2):
    A = []
    B = []
    str1 = str1.lower()
    str2 = str2.lower()
    
    for i in range(len(str1)-1):
        if str1[i].isalpha() and str1[i+1].isalpha():
            A.append(str1[i]+str1[i+1])
    for i in range(len(str2)-1):
        if str2[i].isalpha() and str2[i+1].isalpha():
            B.append(str2[i]+str2[i+1])
            
    if not len(A) and not len(B):
        return 65536
    
    hap = len(A)+len(B)
    kyo = 0
    
    for a in A:
        for b in B:
            if a == b:
                kyo += 1
                B.remove(a)
                hap -= 1
                break
    
    return int(kyo/hap*65536)
